docs/content/en/docs/Design/secret.md

```diff
@@ -0,0 +1,249 @@
+  } -> !secret.secret<tensor<8xi16>>
+  %6 = secret.generic ins(%4, %5 : !secret.secret<tensor<8xi16>>, !secret.secret<tensor<8xi16>>) {
+  ^bb0(%arg2: tensor<8xi16>, %arg3: tensor<8xi16>):
+    %8 = arith.addi %arg2, %arg3 : tensor<8xi16>
+    secret.yield %8 : tensor<8xi16>
+  } -> !secret.secret<tensor<8xi16>>
+  %7 = secret.generic ins(%6 : !secret.secret<tensor<8xi16>>) {
+  ^bb0(%arg2: tensor<8xi16>):
+    %extracted = tensor.extract %arg2[%c7] : tensor<8xi16>
+    secret.yield %extracted : i16
+  } -> !secret.secret<i16>
+  return %7 : !secret.secret<i16>
+}
+```
+
+And then lowering it to `bgv` with `--secret-to-bgv="poly-mod-degree=8"` (the
+pass option matches the tensor size, but it is an unrealistic FHE polynomial
+degree used here just for demonstration purposes). Note type annotations on ops
+are omitted for brevity.
+
+```mlir
+#encoding = #lwe.polynomial_evaluation_encoding<cleartext_start = 16, cleartext_bitwidth = 16>
+#params = #lwe.rlwe_params<ring = <cmod=463187969, ideal=#_polynomial.polynomial<1 + x**8>>>
+!ty1 = !lwe.rlwe_ciphertext<encoding=#encoding, rlwe_params=#params, underlying_type=tensor<8xi16>>
+!ty2 = !lwe.rlwe_ciphertext<encoding=#encoding, rlwe_params=#params, underlying_type=i16>
+
+func.func @dot_product(%arg0: !ty1, %arg1: !ty1) -> !ty2 {
+  %c1 = arith.constant 1 : index
+  %c2 = arith.constant 2 : index
+  %c4 = arith.constant 4 : index
+  %c7 = arith.constant 7 : index
+  %0 = bgv.mul %arg0, %arg1
+  %1 = bgv.relinearize %0 {from_basis = array<i32: 0, 1, 2>, to_basis = array<i32: 0, 1>}
+  %2 = bgv.rotate %1, %c4
+  %3 = bgv.add %1, %2
+  %4 = bgv.rotate %3, %c2
+  %5 = bgv.add %3, %4
+  %6 = bgv.rotate %5, %c1
+  %7 = bgv.add %5, %6
+  %8 = bgv.extract %7, %c7
+  return %8
+}
+```
+
+## Differences for CGGI-style pipeline
+
+The `tosa-to-boolean-tfhe` and related pipelines add a few additional steps.
+The main goal here is to apply a hardware circuit optimizer to blocks of standard MLIR code
+(inside `secret.generic` ops)
+which converts the computation to an optimized boolean circuit with a desired set of gates.
+Only then is `-secret-distribute-generic` applied to split the ops up
+and lower them to the `cggi` dialect.
+In particular, because passing an IR through the circuit optimizer
+requires unrolling all loops,
+one useful thing you might want to do is to optimize only the _body_ of a for loop nest.
+
+To accomplish this, we have two additional mechanisms.
+One is the pass option `ops-to-distribute` for `-secret-distribute-generic`,
+which allows the user to specify a list of ops that `generic` should be split across,
+and all others left alone. Specifying `affine.for` here will pass `generic` through
+the `affine.for` loop, but leave its body intact.
+This can also be used with the `-unroll-factor` option to the `-yosys-optimizer` pass
+to partially unroll a loop nest and pass the partially-unrolled body through
+the circuit optimizer.
+
+The other mechanism is the `secret.separator` op,
```

Comment:
Another key reason: this allows us to insert boundaries between higher-level units of the program, before lowering those down and requiring an op separator belonging to a lower level dialect. We can save this for the TOSA example.

Reply:
Ah you mention it below!